.github/workflows/build-test-ship.yaml

```diff
@@ -183,34 +197,248 @@ jobs:
             ${{ env.ECR_REGISTRY }}/${{ env.ECR_REPOSITORY }}:${{ env.IMAGE_TAG }}
             ${{ env.ECR_REGISTRY }}/${{ env.ECR_REPOSITORY }}:latest
 
+  build-for-target-platforms:
+    name: Verify ${{ matrix.service }} images can be built on target platforms
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        service: ${{ fromJson(inputs.services) }}
+    if: github.event_name == 'pull_request' && !github.event.pull_request.draft
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build images on target platforms
+        uses: docker/build-push-action@v3
+        with:
+          context: .
+          file: ./cmd/${{ matrix.service }}/Dockerfile
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          platforms: linux/amd64,linux/arm64
+
+  run-api-proctor-tests:
+    name: Run API Proctor tests
+    if: inputs.run-api-proctor-tests
+    runs-on: ubuntu-latest
+    needs: build-and-ship-images
+    timeout-minutes: ${{ inputs.api-proctor-timeout-minutes }}
+    env:
+      PGADMIN_PORT: 5000
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_SERVER: postgres
+      POSTGRES_PORT: 5432
+    steps:
+      - name: Checkout development repo
+        uses: actions/checkout@v3
+        with:
+          repository: nicheinc/development
+          token: ${{ secrets.NPM_WRITE_TOKEN }}
+          submodules: true
+
+      - name: Download docker image artifacts
+        uses: actions/download-artifact@v3
+
+      - name: Set up yq for YAML processing
+        uses: mikefarah/yq@v4.27.3
+
+      - name: Get relevant services as a space-separated string
+        id: get-services
+        shell: bash
+        run: |
+          tools=`yq -o=json '.services
+            | with_entries(select(.value.profiles.[] | contains("tools")))
+            | to_entries
+            | map(.key)' docker-compose.yml`
+          all_services=`yq -o=json '.services | keys' docker-compose.yml`
+          non_tools=`jq --null-input "${all_services} - ${tools}"`
+          input_services='${{ inputs.services }}'
+          services=`jq --null-input "${input_services} - (${input_services} - ${non_tools})"`
```

Comment:
D'oh, of course we'll need the dependencies of the dependencies, and I see we traverse the graph later.

Also after more reading, I think I now understand that the issue is that service repos include tools, and we want to exclude tools, so `input_services` itself includes tools, in which case my new expression is not useful. However, I still think `input_services - tools` might be sufficient for what we have going on here at the moment, and more clear.

Reply:
**tl;dr: I don't think we can just do `input_services - tools` because not every "tool" in `inputs.services` is in `tools`.**

Sorry this algorithm and my explanation are confusing. 🙈 There are a few different sets of executables we're dealing with here:
- The `services` input the current repo passes to `build-test-ship`, which includes all executables that need to be built and shipped to ECR, including:
  - (A) Actual services (microservices, Kafka consumers, etc.)
  - (B) Non-service command-line tools
- Compose file services, including those that are:
  - (C) Not part of the `tools` profile
  - (D) Part of the `tools` profile

We're given A ∪ B (via `inputs.services`), C ∪ D (via `yq`), and D (via `yq`), and what we really want is A: just the services that (along with transitive dependencies) need to be running before we start the API Proctor tests. We also know that:
- A and B are disjoint (by definition)
- C and D are disjoint (by defintion)
- B and C are disjoint (since any tools that _are_ included are in fact marked with the `tools` profile)
- A ⊆ C (empirically, thankfully)

You'd think B ⊆ D, but unfortunately some executables are missing from the Compose file entirely. For instance, `lead` [passes `lead-scheduler`](https://github.com/nicheinc/lead/blob/801e256819ff0931a104c3c9caa4d0d55c796ad7/.github/workflows/ci.yaml#L32), which is absent from the [`docker-compose.yml`](https://github.com/nicheinc/development/blob/master/docker-compose.yml) in `development`. (Maybe we should add it there? But that's not the state of the world today, and there are about five other repos in a similar situation.)

So instead I think we have to do:
1. A = (A ∪ B) ∩ C, since A ⊆ C and B/C disjoint
2. C = (C ∪ D) - D, since they're disjoint
3. A = (A ∪ B) ∩ ((C ∪ D) - D), by (1) and (2)
4. A = A ∪ B - (A ∪ B - ((C ∪ D) - D)), by [math](https://proofwiki.org/wiki/Set_Difference_with_Set_Difference#:~:text=3%20Proof%202-,Theorem,T%E2%88%96(T%E2%88%96S))

That last step is necessary because there's no intersection operation built into `jq`.

Reply:
Thank you for the explanation! It makes sense to me now.